### DISCUSSION TITLE
Gphoto2 need to adjust the     libgphoto2/camlibs  /ptp2/library.c file

### DISCUSSION THREAD

**Discussion by Striker62** [photobooth-app/photobooth-app#199] [Q&A] [answered]
Hi, I used gphoto2 same years ago with my Nikon D3100 in PTP mode. After days of research i found a hint to comment the Nikon D3100 section in the  libgphoto2/camlibs  /ptp2/library.c file . That worked perfect.
I'm not able to find this file on my raspberry.
Can you please help.

**Comment by mgineer85** [comment]
If you want to go that way you need the sourcecode of libgphoto and install the python bindings to use the compiled version. I recommend not to go that way usually.
Just editing a .c file would have no effect…
What are the issues you run into? Maybe different mode of the camera helps?

**Comment by Striker62** [comment]
According to the list of suooprted Cameras my Nikon D3100 should work.
Years ago I build a solution to manage the Camera ; Take Piktures , Download Files ....
It was also on a Raspberry pi . But I thing it was not Python.

At the beginning the Camera was not detected. I did a Firmware update on the Camera. Then the Camera was detected in PTP mode but gphoto2 show some errors.

Finaly i found this and it worked.
But if here we have another implementation I have to find another solution or use another camera.

https://github.com/gphoto/libgphoto2/issues/140
>>>>I think I figured out the problem after mucking around a bit. The D3000/D3100 don't seem to support some of the operations that the camera library seems to think they do. I had to clone the library and build/install it myself from source since this requires modifying some of the source files.

All I did was look in the debug output and looked for lines with "PTP Operation Not Supported" and commented out the operations that errored out.

Comment out the following lines in the file libgphoto2/camlibs/ptp2/library.c:

(Note the change in indicies and the realloc size)

if ((nikond >= 3000) && (nikond < 3199)) {
  GP_LOG_D("The D3xxx series hides commands from us ... adding all D3000 ones");
  if (!ptp_operation_issupported(&camera->pl->params, PTP_OC_NIKON_GetVendorPropCodes)) {
    C_MEM (di->OperationsSupported = realloc(di->OperationsSupported,sizeof(di->OperationsSupported[0])*(di->OperationsSupported_len + 8)));
    //di->OperationsSupported[di->OperationsSupported_len+0]  = PTP_OC_NIKON_GetVendorPropCodes;
    //di->OperationsSupported[di->OperationsSupported_len+1]  = PTP_OC_NIKON_CheckEvent;
    //di->OperationsSupported[di->OperationsSupported_len+2]  = PTP_OC_NIKON_Capture;
    di->OperationsSupported[di->OperationsSupported_len+0]  = PTP_OC_NIKON_AfDrive;
    //di->OperationsSupported[di->OperationsSupported_len+4]  = PTP_OC_NIKON_SetControlMode;
    di->OperationsSupported[di->OperationsSupported_len+1]  = PTP_OC_NIKON_DeviceReady;
    //di->OperationsSupported[di->OperationsSupported_len+6]  = PTP_OC_NIKON_AfCaptureSDRAM;
    di->OperationsSupported[di->OperationsSupported_len+2]  = PTP_OC_NIKON_DelImageSDRAM;

    di->OperationsSupported[di->OperationsSupported_len+3]  = PTP_OC_NIKON_GetPreviewImg;
    //di->OperationsSupported[di->OperationsSupported_len+9]  = PTP_OC_NIKON_StartLiveView;
    //di->OperationsSupported[di->OperationsSupported_len+10] = PTP_OC_NIKON_EndLiveView;
    //di->OperationsSupported[di->OperationsSupported_len+11] = PTP_OC_NIKON_GetLiveViewImg;
    di->OperationsSupported[di->OperationsSupported_len+4] = PTP_OC_NIKON_MfDrive;
    di->OperationsSupported[di->OperationsSupported_len+5] = PTP_OC_NIKON_ChangeAfArea;
    di->OperationsSupported[di->OperationsSupported_len+6] = PTP_OC_NIKON_AfDriveCancel;
    /* only opcodes up to 9206 ... I think. */
    di->OperationsSupported_len += 7;
  }
}

After recompiling and reinstalling, running:

gphoto2 --camera="Nikon DSC D3000 (PTP mode)" --capture-image

no longer produces any errors, and an image is taken. I have not tested the D3100, but the above code seems to affect it as well. Commenting out the above lines has not affected any of the other functionality from what I can tell. <<<<

**Reply by mgineer85** [reply]
Ok, it seems you know what your doing , so following should help you to build your own solution.
We use the pypi package python-gphoto2. For convenience we prefer the installation including the precompiled libgphoto in the wheel.

What you want is a local installation, described here: https://github.com/jim-easterbrook/python-gphoto2/blob/main/INSTALL.rst#local-libgphoto2

to use the local installation you need to change the photobooth installation command to prefer the local installation instead the precompiled binary.

this installer might be helpful for you also as install instructions for gphoto can be complicated https://github.com/gonzalo/gphoto2-updater

**Reply by mgineer85** [reply]
Another note: the problem seems to be fixed in gphoto according to the linked issue.
What actual problems do you have. Can you provide logs? Liveview is not supported by the camera, but pictures should be fine…

**Comment by Striker62** [comment]
"Ok, it seems you know what your doing , "  ;-) not realy I found a solution i tried to follow  (not realy know the background). it Worked 
Will send some log files . Thanks

**Comment by Striker62** [comment]
find here the Photobooth-app log file
[photobooth_20240329.log](https://github.com/photobooth-app/photobooth-app/files/14802488/photobooth_20240329.log)
as you can see the camera is correct detected but but with capture image option.

**Comment by Striker62** [comment]
find here the gphoto debug from capture-image
[photo2-logfile.txt](https://github.com/photobooth-app/photobooth-app/files/14802665/photo2-logfile.txt)

**Reply by mgineer85** [reply]
Where did you get that logs from? Did never see it before… I was thinking about the PhotoBooth logs.

**Reply by Striker62** [reply]
i send 2 files in 2 posts in the first you have the app log the second is fron gphoto2

**Reply by mgineer85** [reply]
Sorry, missed that.
For Nikon please disable the viewfinder setting in config backends.
Also Nikon seems to take much time in transfer, but we only wait 4 seconds, which could be too low and abort too early.
Please check this issue if it applies here also https://github.com/photobooth-app/photobooth-app/issues/176 try to lower the quality, no raw shooting and the you will find out if it’s working in general, afterwards the finetuning.

**Reply by Striker62** [reply]
the issue is solvled :-)  
you pointed me in the right direction.
It was not the sizze of the pictures, it was the autofocus that took too much time. 

Now I can Start realy with play around with all of you development :-)
First will be the frames .

**Reply by mgineer85** [reply]
Ok, cool! What did you do to reduce times? How long does it take to capture and transfer the images? The shorter, the better the user experience 👍

**Reply by Striker62** [reply, marked as answer]
I deactivated auto focus . It takes 1 to 2 sec to capture and download the picture.

What I see ist the the count down for the trigger ( the cricle when you click take a picture ) runs from 2 sec and the camera take the picture at 0.7 sec instead of 0.0. 

my gues is that the trigger start before the webpage is starting to show the count down.

Any Idea to get it at 0.0. ?

**Reply by Striker62** [reply]
ok I found the parameter in the Config but I set it to 0 and now the trigger is at 0.5 sec
any other parameter I missed

**Reply by mgineer85** [reply]
right the setting countdown camera offset in config->common is used to offset the camera delay. so 0.5s sound reasonable for your case.
There is also config->uisettings where you can adjust the time of the "smile" message which you might want to raise.
Just play around with these values and hopefully you find something that works well for you.

**Comment by Striker62** [comment]
any thing else that help to find the issue ?

**Reply by mgineer85** [reply]
Yes, PhotoBooth logs would be good. But I think you run into another issue currently, please check this one https://photobooth-app.org/support/faq/#gphoto2-camera-found-but-no-access

**Reply by Striker62** [reply]
nop, i found this and remove the service . I also tried with and with out sd card. all logs i send was done after remove of the service.
